hawc/apps/animalv2/forms.py

```diff
@@ -0,0 +1,356 @@
+    class Meta:
+        model = models.AnimalGroup
+        exclude = ("experiment",)
+
+    def __init__(self, *args, **kwargs):
+        experiment = kwargs.pop("parent", None)
+        super().__init__(*args, **kwargs)
+        if experiment:
+            self.instance.experiment = experiment
+
+    @property
+    def helper(self):
+        helper = BaseFormHelper(self)
+        helper.form_tag = False
+        helper.add_row("species", 3, "col-md-4")
+        helper.add_row("lifestage_at_exposure", 2, "col-md-6")
+        helper.add_row("generation", 2, "col-md-6")
+        helper.add_row("husbandry_and_diet", 2, "col-md-6")
+        set_textarea_height(self.fields)
+
+        assessment_id = self.instance.experiment.study.assessment.pk
+        helper.add_create_btn(
+            "species",
+            reverse("assessment:species_create", args=(assessment_id,)),
+            "Create species",
+        )
+        helper.add_create_btn(
+            "strain",
+            reverse("assessment:strain_create", args=(assessment_id,)),
+            "Create strain",
+        )
+        return helper
+
+    def is_valid(self):
+        if "species" in self.data and "strain" in self.data:
+            species_obj = assessment_models.Species.objects.get(pk=self.data["species"])
+            strain_obj = assessment_models.Strain.objects.get(pk=self.data["strain"])
+            if species_obj.id != strain_obj.species.id:
+                self.add_error(
+                    "strain",
+                    "Strain must be of the same species as the selected species",
+                )
+        return super().is_valid()
+
+
+class TreatmentForm(forms.ModelForm):
+    class Meta:
+        model = models.Treatment
+        exclude = ("experiment",)
+
+    def __init__(self, *args, **kwargs):
+        experiment = kwargs.pop("parent", None)
+        super().__init__(*args, **kwargs)
+        if experiment:
+            self.instance.experiment = experiment
+        self.fields["chemical"].queryset = self.instance.experiment.v2_chemicals.all()
+
+    @property
+    def helper(self):
+        helper = BaseFormHelper(self)
+        helper.form_tag = False
+        helper.add_row("name", 3, "col-md-4")
+        helper.add_row("exposure_duration", 3, "col-md-4")
+        set_textarea_height(self.fields)
+
+        return helper
+
+
+class DoseGroupForm(forms.ModelForm):
+    formset_parent_key = "treatment_id"
+
+    class Meta:
+        model = models.DoseGroup
+        exclude = ("treatment",)
+
+    def __init__(self, *args, **kwargs):
+        treatment = kwargs.pop("parent", None)
+        super().__init__(*args, **kwargs)
+        if treatment:
+            self.instance.treatment = treatment
+
+
+# thx https://stackoverflow.com/questions/21754918/rendering-tabular-rows-with-formset-in-django-crispy-forms and https://stackoverflow.com/questions/42615357/cannot-pass-helper-to-django-crispy-formset-in-template
+class DoseGroupFormHelper(BaseFormHelper):
```

Comment:
make a subclass in common.forms? and use here and elsewhere in this file.

Reply:
done, added FormsetGenericFormHelper in common.forms